proposals/0724-qualified-lists.rst

```diff
@@ -0,0 +1,395 @@
+  pattern FromListNil = HNil
+
+  {-# COMPLETE FromListCons, FromListNil #-}
+
+Users could then do
+
+::
+
+  import Data.HList.Qualified qualified as HList
+
+  HList.[Just True, Just 1, Nothing] :: HList Maybe '[Bool, Int, String]
+
+  -- hlist matches all the patterns below
+  case hlist of
+    HList.[Just True, _, Nothing] -> _
+    HList.[_, Just 1, Nothing] -> _
+    Just _ HList.: _ -> _
+
+ByteArray
+~~~~~~~~~
+
+Example of a ``ByteArray`` implementation, which requires knowing the length of the list in advance.
+
+::
+
+  type Builder s = Int -> MutableByteArray s -> ST s ()
+
+  buildList ::
+    forall a. Prim a =>
+    Integer ->
+    ( forall s.
+      (a -> Builder s -> Builder s)
+      -> Builder s
+      -> Builder s
+    ) ->
+    ByteArray
+  buildList n f = createByteArray (n * sizeOfType @a) $ f cons nil
+    where
+      nil :: Builder s
+      nil = \_ _ -> pure ()
+
+      cons :: Prim a => a -> Builder s -> Builder s
+      cons x next = \i arr -> writeByteArray arr i x >> next (i + 1) arr
+
+  -- [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf0, 0x3f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40]
+  print ByteArray.[1, 2]
+
+Effect and Interactions
+-----------------------
+
+With ``QualifiedLists``, there's no more typeclass ambiguity; e.g. if the ``vector`` library provided ``Data.Vector.Qualified`` as described in *Section 4.1 Vector*, users can do
+
+::
+
+  import Data.Vector.Qualified qualified as V
+
+  main = print V.[False, True]
+
+The equivalent code with ``OverloadedLists`` would have failed to compile with ``Couldn't match expected type 'Item a0' with actual type 'Bool'``.
+
+
+Interactions with other extensions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Related to `QualifiedStrings <https://github.com/ghc-proposals/ghc-proposals/pull/723>`_ and `QualifiedNumerics <https://github.com/ghc-proposals/ghc-proposals/pull/725>`_, but all three proposals are orthogonal to each other.
+
+* `Allow arbitrary identifiers as fields in OverloadedRecordDot <https://github.com/ghc-proposals/ghc-proposals/pull/668>`_ has similar syntax to the proposed qualified string literal, but as ``M.bar`` is parsed as a qualified identifier even with OverloadedRecordDot, it makes sense that ``M."bar"`` is also parsed as a qualified literal.
+
+Costs and Drawbacks
+-------------------
+
+Development and maintenance should be low effort, as the core implementation is in the renamer step, and typechecking would proceed as normal.
+
+The syntax is approachable for novice users and shouldn't be an extra barrier to understand.
+
+Backward Compatibility
+----------------------
+
+No breakage, as the new syntax is only enabled with the extension.
+
+Furthermore, turning on the extension will generally not break existing code, as the expression would be parsed as function composition between a data constructor and a literal, which would only typecheck if someone adds an ``IsList`` instance for a function type.
```

Comment:
this reads a bit confusingly because it sounds like you're saying "if you turn on the extension, the expression would be parsed as function composition", but I think you mean that's how it would be parsed *before* turning on the extension, right?

Reply:
Yup, rewording